crmsh/corosync.py

```diff
@@ -669,215 +649,370 @@ def remove_qdevice_db(self):
+
+                new_line = False
+                if prev_content_line and prev_content_line != content_line:
+                    new_line = True
+
+                # current token is: {
+                if token_type == LBRACE:
+                    key_prefix_list.append(prev_token_string)
+                    key_prefix = '.'.join(key_prefix_list)
+                    section_key = self._unused_inner_key(key_prefix)
+                    section_key_list.append(section_key)
+                    self._config_dict[section_key] = "{"
+
+                # restore prev line into dict
+                if new_line and ':' in prev_content_line:
+                    key, *value = prev_content_line.split(':')
+                    key = "{}.{}".format(key_prefix, key.strip())
+                    self._config_dict[self._unused_inner_key(key)] = ':'.join(value).strip()
+
+                # current token is: }
+                if token_type == RBRACE:
+                    section_key = section_key_list.pop()
+                    self._config_dict[section_key+"__END"] = "}"
+                    key_prefix_list.pop()
+                    key_prefix = '.'.join(key_prefix_list)
+
+                prev_token_string = token_string
+                prev_content_line = content_line
+
+    def get(self, path, index=0):
+        """
+        Gets the value for the path
+        path: config path
+        index: known index in section
+        """
+        key = self._inner_key(path, index)
+        if key not in self._config_dict or self._is_inner_section(key):
+            return None
+        return self._config_dict[key]
+
+    def get_all(self, path):
+        """
+        Returns all values matching path
+        """
         ret = []
-        for t in self._tokens:
-            if t.token == _tVALUE:
-                ret.append(t.path)
+        for key, value in self._config_dict.items():
+            if self._matched_inner_key(path, key) and not self._is_inner_section(key):
+                ret.append(value)
         return ret
-
+        
     def count(self, path):
-        """Returns the number of elements matching path"""
+        """
+        Returns the number of elements matching path
+        """
         n = 0
-        for t in self._tokens:
-            if t.path == path:
+        for key, value in self._config_dict.items():
+            if self._matched_inner_key(path, key) and value != "}":
                 n += 1
         return n
 
-    def remove(self, path):
-        """Removes the given section or value"""
-        i, e = self.find_bounds(path)
-        if i < 0:
-            return
-        self._tokens = self._tokens[:i] + self._tokens[(e+1):]
+    @contextmanager
+    def _operate_config_list(self):
+        """
+        The contextmanager for convert between config dict and list
+        """
+        self._config_list = list(self._config_dict.items())
+        yield
+        self._config_dict = dict(self._config_list)
```

Comment:
You should be careful here. If you have a context manager, anything that is executed in the with block can raise an exception. If an exception is raised, you need to make sure some "teardown" code is executed properly.

The safest way is to use a `try... finally` construct:

```suggestion
        self._config_list = list(self._config_dict.items())
        try:
            yield
        finally:
            self._config_dict = dict(self._config_list)
```

Maybe it's not really needed, but this is usually the structure how a contextmanager should look like. See the example in [contextlib.contextmanager](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager).

Reply:
Thanks, changed!